08openmp/index.md

```diff
@@ -0,0 +1,307 @@
+    sum += vec[i];
+  }
+
+  double elapsed = timer.elapsed();
+
+  std::cout << "Time: " << elapsed << '\n';
+  std::cout << "Result: " << M_PI*sum/N << '\n';
+```
+
+Compiling and running without OpenMP gives:
+
+```
+Time: 0.276116
+Result: 2
+```
+
+This is exactly what we expect. If you enable OpenMP however, you'll probably get something like:
+
+```
+Time: 0.957646
+Result: 0.14947
+```
+
+I recommend you take a break from reading/coding here and think about the questions, *why has the time increased with more threads* and *why is the answer wrong*? Don't worry if it's not obvious, this is a subtle and tricky bug and requires thinking about what each thread is doing as it accesses each variable inside the loop.
+
+---
+
+This behaviour is the result of a **data race**. Consider even just two threads working in parallel inside the loop. Thread 1 reads data from `vec[i]`. The `for` construct has split the loop between the two threads so its value of `i` is totally unique to thread 1 and it is the only thread touching the data in `vec[i]`. It then needs to add this value into `sum`. It reads `sum`, adds `vec[i]` to it and writes it back into `sum`. This addition takes some time however, and thread 2 is performing the exact same operation but using a different value of `i`. When thread 1 reads `sum` and is currently spending time adding, thread 2 could easily have just finished an addition and has written a new value to `sum`. Thread 1 has no idea that thread 2 has done this and then unknowingly *overwrites* the updated value of `sum` with its own version. Not only does this mean the value of `sum` is *incorrect* but threads are not allowed to write to the same variable at the same time, so each threads must wait for the other to finish, adding more time to the overall calculation. This is why the overall time has increased, despite using multiple threads, and why the result is wrong!
+
+So how do we handle a data race? Each example of a data race is unique and must be handled in its own way, but the general solution is to limit threads to writing to variables no other threads will write to. We'll look later at a few different ways to tackle this particular problem but summing is part of a larger group of operations called **reductions**, where a large data structure (like a vector) is reduced to a smaller number of values, in this case just a single number. Thankfully, OpenMP provides a way to easily deal with these kinds of operations: reduction clauses.
+
+### Example 3: Summing an array with `reduction`
+
+OpenMP allows constructs to be augmented with **clauses**, extra pieces of information we can add to its directives to further refine the parallelisation. In the above sum example, we can tell OpenMP we're trying to perform a reduction by adding the `reduction` clause to the `parallel for` construct along with the operator we want to reduce using and the variable that is holding the value of the reduction:
```

Comment:
A bit more detail on this syntax would be good: what operators can they use? Can they define their own by passing in functions? Can they use lambda functions? 

Also whilst telling them that OpenMP can solve the problem using reductions, it would be valuable to tell them an idea of _how_ the problem is actually solved so they understand what is actually going on.